templates/CRM/Contact/Form/Relationship.tpl

```diff
@@ -144,10 +144,113 @@
       CRM.$(function($) {
         var
           $form = $("form.{/literal}{$form.formClass}{literal}"),
-          relationshipData = {/literal}{$relationshipData|@json_encode}{literal};
-        $('[name=relationship_type_id]', $form).change(function() {
+          $relationshipTypeSelect = $('[name=relationship_type_id]', $form),
+          relationshipData = {},
+          contactTypes = {};
+
+        $('body').on('crmOptionsEdited', 'a.crm-option-edit-link', refreshRelationshipData);
+
+        // Initial load and trigger change on select
+        refreshRelationshipData().done(function() {
+          $relationshipTypeSelect.change();
+        });
+
+        /**
+         * Fetch contact types and reset relationship data
+         */
+        function refreshRelationshipData() {
+          var defer = $.Deferred();
+
+          // reset
+          relationshipData = {};
+
+          getContactTypes().then(function() {
+            getRelationshipData().then(function() {
+              defer.resolve();
+            });
+          });
+
+          return defer.promise();
+        }
+
+        /**
+         * Fetches the relationship data using latest relationship types
+         */
+        function getRelationshipData() {
+          var subtype,
```

Comment:
These variables are used only inside $.each, so we can define them inside only. Overall this is how it should look like
```javascript
function getRelationshipData() {
  var defer = $.Deferred();

  if (!$.isEmptyObject(relationshipData)) {
    defer.resolve(relationshipData);
  }

  CRM.api3("RelationshipType", "get")
    .done(function (data) {
      $.each(data.values, function (key, relType) {
        // Please add a comment mentioning what is ["a", "b"]
        $.each(["a", "b"], function (index, suffix) {
          var subtype = relType["contact_subtype_" + suffix];
          var type = subtype || relType["contact_type_" + suffix];
          var label = getContactTypeLabel(type) || "Contact";
          var placeholder = "- select " + label.toLowerCase() + " -";
          relType["placeholder_" + suffix] = placeholder;
        });

        relationshipData[relType["id"]] = relType;
      });

      defer.resolve(relationshipData);
    });

  return defer.promise();
}
```

Reply:
nice, I like the early return!